### PR TITLE
[Hackathon] Basic OpenACC -> OpenMP migration.

### PR DESCRIPTION
**Pull request by olupton**
**Description**
Migrate some offloaded kernels from OpenACC to OpenMP.

The idea with the macros defined here is that the macros `nrn_acc_pragma` and `nrn_omp_pragma` are used to annotate expressions that can be used with either OpenACC **or** OpenMP. i.e.
```c++
nrn_acc_pragma(atomic update)
nrn_omp_pragma(atomic update)
rhs += p * rhs;
```
will expand to **either**
```c++
#pragma acc atomic update
rhs += p * rhs;
```
**or**
```c++
#pragma omp atomic update
rhs += p * rhs;
```
Some other directives may be needed, for example if we want OpenACC directives in a build that prefers OpenMP offload then we may need code like
```c++
#ifdef CORENRN_PREFER_OPENMP_OFFLOAD
// Make sure OpenACC work is done before we start OpenMP work
if (nt->compute_gpu) {
    _Pragma("acc wait(nt->stream_id)")
}
#endif
```

**Use certain branches for the GitLab/SimulationStack CI**
CI_BRANCHES:NMODL_BRANCH=hackathon_main,NEURON_BRANCH=master,

### CODE REVIEW

File: coreneuron/gpu/nrn_acc_manager.cpp [hunk 1]
```diff
@@ -894,67 +905,12 @@ void update_weights_from_gpu(NrnThread* threads, int nthreads) {
...
-        #pragma acc update device(v[0 : ne]) async(_nt->stream_id)
-        #pragma acc update device(rhs[0 : ne]) async(_nt->stream_id)
-        #pragma acc wait(_nt->stream_id)
-        // clang-format on
-    }
-#else
-    (void) _nt;
-#endif
-}
-
```
Comment (alkino) [inline coreneuron/gpu/nrn_acc_manager.cpp:957]: Should we merge that on master as those both function seems unused today?

File: coreneuron/permute/cellorder.cpp [hunk 2]
```diff
@@ -656,28 +650,23 @@ void solve_interleaved1(int ith) {
...
-    #pragma acc parallel loop present(              \
-        nt[0:1], stride[0:nstride],                 \
-        firstnode[0:ncell], lastnode[0:ncell],      \
-        cellsize[0:ncell]) if (nt->compute_gpu)     \
-        async(stream_id)
-// clang-format on
-#endif
+    // OL211123: can we preserve the error checking behaviour of OpenACC's
+    // present clause with OpenMP? It is a bug if these data are not present,
+    // so diagnostics are helpful...
```
Comment (pramodk) [inline coreneuron/permute/cellorder.cpp:656]: // nitpick/note: this we could remove from here, we will post question to mentors.

File: coreneuron/sim/fadvance_core.cpp [hunk 3]
```diff
@@ -78,11 +78,11 @@ void dt2thread(double adt) { /* copied from nrnoc/fadvance.c */
...
-            // clang-format off
-
-            #pragma acc update device(nt->_t, nt->_dt, nt->cj) \
-                    async(nt->stream_id) if(nt->compute_gpu)
-            // clang-format on
+            nrn_pragma_acc(update device(nt->_t, nt->_dt, nt->cj)
+                               async(nt->stream_id) if (nt->compute_gpu))
+            nrn_pragma_omp(target update to(nt->_t, nt->_dt, nt->cj) depend(inout
+                                                                            : nt)
+                               nowait if (nt->compute_gpu))
```
Comment (pramodk) [inline coreneuron/sim/fadvance_core.cpp:85]: nitpick: we should keep clang-format and make these pragmas more readable.

**Review by pramodk — COMMENTED** [review]
Overall looks good to me. As discussed offline, we will fix the clang-formatted part of pragmas manually.

File: CMake/OpenAccHelper.cmake [hunk 4]
```diff
@@ -55,14 +55,20 @@ if(CORENRN_ENABLE_GPU)
...
   # code. Otherwise there may be confusing linker errors. We cannot rely on nvcc and nvc++ using the
   # same default compute capabilities as each other, particularly on GPU-less build machines.
   foreach(compute_capability ${CMAKE_CUDA_ARCHITECTURES})
     string(APPEND NVHPC_ACC_COMP_FLAGS ",cc${compute_capability}")
   endforeach()
+  if(CORENRN_ENABLE_OPENMP AND CORENRN_ENABLE_OPENMP_OFFLOAD)
+    # Enable OpenMP target offload to GPU and if both OpenACC and OpenMP directives are available
+    # for a region then prefer OpenMP.
+    add_compile_definitions(CORENRN_PREFER_OPENMP_OFFLOAD)
+    string(APPEND NVHPC_ACC_COMP_FLAGS " -mp=gpu -Minfo=mp")
```
[the remark this answers is not in the document]
Comment (pramodk) [reply CMake/OpenAccHelper.cmake:70]: this is old comment, I have checked this and this is working fine.

File: coreneuron/sim/fadvance_core.cpp (hunk 3 above)
Comment (pramodk) [reply coreneuron/sim/fadvance_core.cpp:85]: I left this as it is. we will revisit when we will enable clang-format CI check.

File: coreneuron/gpu/nrn_acc_manager.cpp [hunk 5]
```diff
@@ -605,25 +608,33 @@ void update_net_receive_buffer(NrnThread* nt) {
...
-                acc_update_device(nrb->_pnt_index, sizeof(int) * nrb->_cnt);
-                acc_update_device(nrb->_weight_index, sizeof(int) * nrb->_cnt);
-                acc_update_device(nrb->_nrb_t, sizeof(double) * nrb->_cnt);
-                acc_update_device(nrb->_nrb_flag, sizeof(double) * nrb->_cnt);
-                acc_update_device(nrb->_displ, sizeof(int) * (nrb->_displ_cnt + 1));
-                acc_update_device(nrb->_nrb_index, sizeof(int) * nrb->_cnt);
+                nrn_pragma_acc(update device(nrb->_cnt,
+                                             nrb->_displ_cnt,
+                                             nrb->_pnt_index[:nrb->_cnt], nrb->_weight_index
+                                             [:nrb->_cnt], nrb->_nrb_t
```
[the remark this answers is not in the document]
Comment (olupton) [reply coreneuron/gpu/nrn_acc_manager.cpp]: I fixed up some of these manually, although I generally prefer to let `clang-format` do its thing unless it's **really** bad.

File: coreneuron/io/lfp.cpp [hunk 6]
```diff
@@ -8,7 +8,7 @@
 
 namespace coreneuron {
 // extern variables require acc declare
-#pragma acc declare create(pi)
+// #pragma acc declare create(pi)
```
[the remark this answers is not in the document]
Comment (olupton) [reply coreneuron/io/lfp.cpp]: Removed in 587ff3778f947cb43029edc8fffe56a5bbf6cfa9

File: coreneuron/mechanism/eion.cpp [hunk 7]
```diff
@@ -312,7 +294,13 @@ void nrn_init_ion(NrnThread* nt, Memb_list* ml, int type) {
     int _cntml_padded = ml->_nodecount_padded;
     pd = ml->data;
     ppd = ml->pdata;
-    _PRAGMA_FOR_INIT_ACC_LOOP_
+    // No async(...) / nowait / depend clauses.
```
[the remark this answers is not in the document]
Comment (olupton) [reply coreneuron/mechanism/eion.cpp]: Resolved in 587ff3778f947cb43029edc8fffe56a5bbf6cfa9.

File: coreneuron/mechanism/register_mech.cpp [hunk 8]
```diff
@@ -21,9 +21,9 @@ namespace coreneuron {
 int secondorder = 0;
 double t, dt, celsius, pi;
 // declare copyin required for correct initialization
-#pragma acc declare copyin(secondorder)
-#pragma acc declare copyin(celsius)
-#pragma acc declare copyin(pi)
+// #pragma acc declare copyin(secondorder)
+// #pragma acc declare copyin(celsius)
+// #pragma acc declare copyin(pi)
```
[the remark this answers is not in the document]
Comment (olupton) [reply coreneuron/mechanism/register_mech.cpp]: Removed in 587ff3778f947cb43029edc8fffe56a5bbf6cfa9.

File: coreneuron/permute/cellorder.cpp (hunk 2 above)
Comment (olupton) [reply coreneuron/permute/cellorder.cpp:656]: Decided to leave this in place to help ensure we reach a conclusion.

File: coreneuron/sim/fadvance_core.cpp (hunk 3 above)
Comment (olupton) [reply coreneuron/sim/fadvance_core.cpp]: I fixed up some of the worst cases.

File: coreneuron/network/partrans.cpp [hunk 9]
```diff
@@ -41,40 +41,39 @@ void nrnmpi_v_transfer() {
     // gather the source values. can be done in parallel
     for (int tid = 0; tid < nrn_nthread; ++tid) {
         auto& ttd = transfer_thread_data_[tid];
-        auto& nt = nrn_threads[tid];
+        auto* nt = &nrn_threads[tid];
```
Comment (pramodk) [inline coreneuron/network/partrans.cpp:44]: just question - I assume this was just for consistency (and not hidden compiler issues), right?

**Review by pramodk — APPROVED** [review]
LGTM. Only my open question is about `seq` part in `cellorder.cpp`. We can merge this and review it later if you like.

File: coreneuron/network/partrans.cpp (hunk 9 above)
Comment (olupton) [reply coreneuron/network/partrans.cpp:44]: I was using `NrnThread* nt` in the `depend(inout: nt)` clauses, and the compiler rejected `depend(inout: &nt)` here IIRC (maybe it should be `*nt` everywhere, to be revisited when we re-add async stuff.

Comment (pramodk) [reply coreneuron/network/partrans.cpp:44]: Ah ok! make sense.

File: coreneuron/permute/cellorder.cpp [hunk 10]
```diff
@@ -535,10 +538,7 @@ static void bksub_interleaved2(NrnThread* nt,
 #if !defined(_OPENACC)
     for (int i = root; i < lastroot; i += 1) {
 #else
-    // clang-format off
-
-    #pragma acc loop seq
-    // clang-format on
+    nrn_pragma_acc(loop seq)
```
[the remark this answers is not in the document]
Comment (olupton) [reply coreneuron/permute/cellorder.cpp:541]: I added a TODO to the wiki page.

**Pull request merged**